backend/src/test/scala/com/softwaremill/adopttapir/starter/StarterServiceITTest.scala

```diff
@@ -1,51 +1,211 @@
+      val service = ServiceUnderTest(details)
+      service.run(List(helloEndpointTest, docsEndpointTest, metricsEndpointTest).flatten)
     }
   }
+}
 
-  private def createStarterService = {
-    val config = StarterConfig(deleteTempFolder = true, tempPrefix = "sbtService", scalaVersion = "2.13.8")
-    new StarterServiceInTests(config, new ProjectTemplateInTests(config))
+case class ServiceUnderTest(details: StarterDetails) {
+  import Setup._
+
+  def run(tests: List[TestFunction]): Unit = {
+    val log = new StringBuilder()
+    Resource
+      .make(generateZipFile(details, log))(zipFile => IO(zipFile.delete()))
+      .use(zipFile =>
+        createTempDirectory()
+          .flatMap { tempDir =>
+            for {
+              _ <- unzipFile(zipFile, tempDir, log)
+              service <- IO.blocking(Service(tempDir))
+            } yield (tempDir, service)
+          }
+          .bracket { case (_, service) =>
+            for {
+              port <- getPortFromService(service, log)
+              _ <- runTests(port, tests, log)
+            } yield ()
+          } { case (tempDir, service) => service.close() >> IO.blocking(tempDir.delete()) }
+          .timeoutAndForget(90.seconds)
+          .onError { e =>
+            Assertions.fail(
+              s"Only the following test steps were finished for configuration '${details.describe}':$log\n${describe(e)}"
+            )
+          }
+      )
+      .unsafeRunSync()
+  }
+
+  private def generateZipFile(details: StarterDetails, log: StringBuilder): IO[BFile] = {
+    for {
+      zipFile <- ZipGenerator.service.generateZipFile(details).map(_.toScala)
+      _ <- logStep(log, "* zip file was generated")
+    } yield zipFile
   }
 
-  private def sbtCompileTest(file: IO[File]): Unit = {
+  private def createTempDirectory(): IO[BFile] = {
     IO.blocking(BFile.newTemporaryDirectory("sbtTesting"))
-      .bracket { tempDir =>
-        for {
-          zipFile <- file.map(_.toScala)
-          _ <- IO.blocking {
-            zipFile.unzipTo(tempDir)
-            zipFile.delete()
-          }
-          _ <- IO.blocking(os.proc("sbt", ";compile ;test").call(cwd = os.Path(tempDir.toJava)).exitCode shouldBe 0)
+  }
+
+  private def unzipFile(zipFile: BFile, tempDir: BFile, log: StringBuilder): IO[Unit] = {
+    IO.blocking(zipFile.unzipTo(tempDir)) >> logStep(log, "* zip file was unzipped")
+  }
 
-        } yield zipFile
+  private def getPortFromService(service: Service, log: StringBuilder): IO[Integer] = {
+    for {
+      port <- service.port
+      _ <- logStep(log, s"* service compiled, tested & started on port $port")
+    } yield port
+  }
 
-      }(tempDir => IO.blocking(tempDir.delete()))
-      .unsafeRunSync()
+  private def runTests(port: Integer, tests: List[TestFunction], log: StringBuilder): IO[Unit] = {
+    tests.map(test => IO.blocking(test(port))).parSequence >> logStep(
```

Comment:
Open question:

I have a strong feeling that TestFunction should be written in form `type TestFunction = Int => IO[Assertion]` 
Just to emphasis that this will do some side effects. 

It should also simplify calling it.

Reply:
Did it but one still has to call `map` on a list so that effect is created 😄 
```
  tests
      .map(_(port))
      .parSequence
  ...
```